google/cloud/storage/emulator/utils/common.py

```diff
@@ -330,3 +338,88 @@ def enforce_patch_override(request):
         and request.headers.get("X-Http-Method-Override", "") != "PATCH"
     ):
         utils.error.notallowed(context=None)
+
+
+def __extract_data(data):
+    if isinstance(data, FlaskResponse):
+        return data.get_data()
+    if isinstance(data, dict):
+        return json.dumps(data)
+    return data
+
+
+def __wrap_in_response_instance(data, to_wrap):
+    if isinstance(data, FlaskResponse):
+        data.set_data(to_wrap)
+        return data
+    return FlaskResponse(to_wrap)
+
+
+def handle_retry_test_instruction(database, request, method):
+    test_id = request.headers.get("x-retry-test-id", None)
+
+    def retry_test_response_handler(data):
+        return data
+
+    if test_id and database.has_instructions_retry_test(test_id, method):
+        next_instruction = database.peek_next_instruction(test_id, method)
+        error_code_matches = utils.common.retry_return_error_code.match(
+            next_instruction
+        )
+        if error_code_matches != None:
+            database.dequeue_next_instruction(test_id, method)
+            items = list(error_code_matches.groups())
+            error_code = items[0]
+            utils.error.generic(
+                "Retry Test: Caused a {}".format(error_code),
+                rest_code=error_code,
+                grpc_code=None,
+                context=None,
+            )
+        retry_connection_matches = utils.common.retry_return_error_connection.match(
+            next_instruction
+        )
+        if retry_connection_matches != None:
+            items = list(retry_connection_matches.groups())
+            # sys.exit(1) retains database state with more than 1 thread
+            if items[0] == "reset-connection":
+                database.dequeue_next_instruction(test_id, method)
+                fd = request.environ["gunicorn.socket"]
+                fd.setsockopt(
+                    socket.SOL_SOCKET, socket.SO_LINGER, struct.pack("ii", 1, 0)
+                )
+                fd.close()
+                sys.exit(1)
+            elif items[0] == "broken-stream":
+
+                def retry_test_response_handler(data):
```

Comment:
This is fairly confusing to me... I think you are declaring a local function so the value returned below is a different one?

Consider instead having two functions and picking what to return:

```py
def handle_retry_test_instruction(database, request, method):
  def simple_handler(data):
    return data

  def streaming_handler(data):
    ...
   return __wrap_in_response_instance(data, streamer())

  handler = simple_handler # Use simple_handler by default
   ...
  elif items[0] == "broken-stream":
    handler = streaming_handler
  ... ... ...
  ..
  return handler
```

Reply:
Addressed in: https://github.com/googleapis/google-cloud-cpp/pull/6611/commits/b996679af727df8380ef25478f7a839446249473

Reply:
One more refactor: https://github.com/googleapis/google-cloud-cpp/pull/6611/commits/e729f58c793acfeda31d52ffd4a662c316564ce6